File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char (isUpper, isAlpha, isSpace)

responseFor :: String -> String
responseFor sentence
  | isNothing  = "Fine. Be that way!"
  | isForcefulQuestion = "Calm down, I know what I'm doing!"
  | isShouting = "Whoa, chill out!"
  | isShoutingGibberish = "Whoa, chill out!"
  | isQuestion = "Sure."
  | otherwise  = "Whatever."
  where areAllUpper   = all isUpper
        isQuestion    = last (filter (not . isSpace) sentence) == '?'
        isForcefulQuestion = isShouting && isQuestion
        isShouting    = saysSomething sentence && areAllUpper (filter isAlpha sentence)
        isShoutingGibberish = areAllUpper sentence
        saysSomething = any isAlpha
        isNothing     = all isSpace sentence

File: go/bob/bob.go
package bob

import (
  "strings"
)

// Possible answers
const (
	nothing = "Fine. Be that way!"
  forcefulQuestion = "Calm down, I know what I'm doing!"
	shouting = "Whoa, chill out!"
	question = "Sure."
	something = "Whatever."
)

// Receives a sentence then answer according with
// sentence pattern
func Hey(remark string) string {
	remark = strings.TrimSpace(remark)

	switch {
	case len(remark) == 0:
		return nothing
	case isShouting(remark) && isQuestion(remark):
		return forcefulQuestion
  case isQuestion(remark):
		return question
  case isShouting(remark):
		return shouting
	default:
		return something
	}
}

func isQuestion(sentence string) bool {
  return strings.HasSuffix(sentence, "?")
}

func isShouting(sentence string) bool {
  return sentence == strings.ToUpper(sentence) && sentence != strings.ToLower(sentence)
}
